### PR TITLE
feat: run / tiltseries s3 data validation (zarr, mrc, mdoc, and metadata checks)

### PR DESCRIPTION
**Pull request by daniel-ji**
Merge after #207.

See the tiltseries & runs sections of https://docs.google.com/document/d/1yMKM0DW9KRhlcYiBGPcR7oW0liGtUew6NAmBhMg5U3w/edit

### CODE REVIEW

File: ingestion_tools/scripts/data_validation/conftest.py [hunk 1]
```diff
@@ -35,35 +36,56 @@ def run_names(bucket: str, dataset: str, run_glob: str) -> List[str]:
     return run_names
 
 
-def voxel_spacings(bucket: str, dataset: str, voxelspacing_glob: str) -> List[float]:
-    """All voxel spacings matching the vozel spacing glob pattern in the dataset."""
+def get_voxel_spacing_files(bucket: str, dataset: str, run_names: List[str], voxelspacing_glob: str) -> List[str]:
     fs: S3Filesystem = FileSystemApi.get_fs_api(mode="s3", force_overwrite=False)
```
Comment (manasaV3) [inline ingestion_tools/scripts/data_validation/conftest.py:40]: nit: Why is this not a parameter?

File: ingestion_tools/scripts/data_validation/conftest.py [hunk 2]
```diff
@@ -35,35 +36,56 @@ def run_names(bucket: str, dataset: str, run_glob: str) -> List[str]:
...
+        )
+
+        # Combine results
+        for result in results:
+            tentatives += result
 
     exclude = ["Images", "dataset_metadata.json"]
 
     for ex in exclude:
         tentatives = [tent for tent in tentatives if ex not in tent]
```
Comment (manasaV3) [inline ingestion_tools/scripts/data_validation/conftest.py:59]: This check should be a part of the fixture that generates the run_names and not be done here. [Which it already is](https://github.com/chanzuckerberg/cryoet-data-portal-backend/blob/1deb4ab6c71c3347ee9b77c5fdcff4b412ce62c8/ingestion_tools/scripts/data_validation/conftest.py#L32-L34).

File: ingestion_tools/scripts/data_validation/conftest.py [hunk 3]
```diff
@@ -35,35 +36,56 @@ def run_names(bucket: str, dataset: str, run_glob: str) -> List[str]:
...
     fs: S3Filesystem = FileSystemApi.get_fs_api(mode="s3", force_overwrite=False)
-    tentatives = fs.glob(f"s3://{bucket}/{dataset}/*/Tomograms/{voxelspacing_glob}")
+
+    tentatives = []
+
+    # Use ThreadPoolExecutor to run glob_task in parallel
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        # Map the run_names to the glob_task function
+        results = executor.map(
+            lambda run_name: fs.glob(f"s3://{bucket}/{dataset}/{run_name}/Tomograms/{voxelspacing_glob}"),
```
Comment (manasaV3) [inline ingestion_tools/scripts/data_validation/conftest.py:48]: nit: could we create a formatted string outside the lambda, which already has the bucket, dataset, voxelspacing_glob applied? So, that this formatting only has to apply the run name?

File: ingestion_tools/scripts/data_validation/tests/helper_mrc_zarr.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    def test_zarr_mrc_volume_size(self):
+        """Check that the zarray shape (volume size) matches the mrc volume size, for each zarr binning."""
+
+        def check_volume_size(header_data, zarr_filename):
+            mrc_file = zarr_filename.replace(".zarr", ".mrc")
+
+            zarrays = header_data["zarrays"]
+            for i, zarray in zarrays.items():
+                header = self.mrc_headers[mrc_file].header
```
Comment (manasaV3) [inline ingestion_tools/scripts/data_validation/tests/helper_mrc_zarr.py:28]: nit: Could this header be outside the for loop?

File: ingestion_tools/scripts/data_validation/tests/test_tiltseries.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        assert len(set(tiltseries_mdoc["PixelSpacing"])) == 1
+        assert tiltseries_metadata["pixel_spacing"] == tiltseries_mdoc["PixelSpacing"][0]
+
+    def test_tiltseries_image_size_mdoc(self, tiltseries_metadata: Dict, tiltseries_mdoc: pd.DataFrame):
+        """Check that the tiltseries image size matches the MDOC data."""
+        assert len(set(tiltseries_mdoc["ImageSize"])) == 1
+        assert tiltseries_metadata["size"]["x"] == tiltseries_mdoc["ImageSize"][0][0]
+        assert tiltseries_metadata["size"]["y"] == tiltseries_mdoc["ImageSize"][0][1]
+
+    ### END MDOC consistency tests ###
```
Comment (manasaV3) [inline ingestion_tools/scripts/data_validation/tests/test_tiltseries.py:122]: While the mdoc files are saved along with the tiltseries at the moment, they are associated to the frames. As, there is processing to go from frames to tiltseries, it is very unlikely for us to have the same Image size or pixel spacings. So, these two tests are more likely to fail and that is the expected behaviour.

**Review by manasaV3 — CHANGES_REQUESTED** [review]

File: ingestion_tools/scripts/data_validation/conftest.py (hunk 1 above)
Comment (daniel-ji) [reply ingestion_tools/scripts/data_validation/conftest.py:40]: fixed in a later pr!

File: ingestion_tools/scripts/data_validation/conftest.py (hunk 2 above)
Comment (daniel-ji) [reply ingestion_tools/scripts/data_validation/conftest.py:59]: fixed in a later pr, see #250

File: ingestion_tools/scripts/data_validation/conftest.py (hunk 3 above)
Comment (daniel-ji) [reply ingestion_tools/scripts/data_validation/conftest.py:48]: done in #250!

File: ingestion_tools/scripts/data_validation/tests/helper_mrc_zarr.py (hunk 4 above)
Comment (daniel-ji) [reply ingestion_tools/scripts/data_validation/tests/helper_mrc_zarr.py:28]: fixed!

File: ingestion_tools/scripts/data_validation/tests/test_tiltseries.py (hunk 5 above)
Comment (daniel-ji) [reply ingestion_tools/scripts/data_validation/tests/test_tiltseries.py:122]: Would it be of interest to see which runs fail against these tests though? Maybe worth asking Utz tomorrow?

**Review by manasaV3 — APPROVED** [review]

**Pull request merged**